### ISSUE TITLE
Dropdown.prototype.toggle() should ignore defaultPrevented click events

### ISSUE THREAD

**Issue by jasonlo87** [twbs/bootstrap#17155]
Ran into a simple looking issue w/ dropdown.js:

1) Catch some event before BS and clear the .open class off a dropdown.
2) PreventDefault on the event in the same handler.
3) Dropdown.prototype.toggle erroneously tries to check for isDefaultPrevented() AFTER it's blasted away the original event and replaced it with a brand new one.

Result:
Bootstrap re-toggles my dropdown back open.

Expected:
Bootstrap respects my preventDefault and doesn't toggle the dropdown back open.

Here's a JSFiddle displaying the issue. It appears that the isDefaultPrevented() check should just be moved up a few lines (it also appears to occur like this in clearMenus() and maybe a few other methods in there): https://jsfiddle.net/9che8vz3/5/

**Comment by twbs-lmvtfy** [comment]
Hi @jasonlo87!

You appear to have posted a live example (https://fiddle.jshell.net/9che8vz3/4/show/light/), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 80, column 1: [E024](https://github.com/twbs/bootlint/wiki/E024): `.panel-heading` must have a `.panel` parent

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by jasonlo87** [comment]
Revised JSFiddle per the bot thing:
https://jsfiddle.net/9che8vz3/5/

**Comment by jasonlo87** [comment]
Revised the fiddle in the initial comment per the bot thing.

**Comment by cvrebert** [comment]
> It appears that the isDefaultPrevented() check should just be moved up a few lines

Actually its current location is deliberate, so that folks can stop the dropdown from showing by preventDefault-ing the `show.bs.dropdown` event.
Your suggestion to add a additional check for whether the `click` event was preventDefault-ed sounds reasonable though.

**Comment by cvrebert** [comment]
Changing this to v4 since it would break backward compatibility.

**Comment by fat** [comment]
as far as i can tell, you're not even listening to a bootstrap custom event… so preventDefault isn't going to do any favors for you in your fiddle.

bootstrap has no way to even know you're click is happening.

our preventdefault that chris is talking about is only available for our custom events… which this isn't using.

So going to close for now, feel free to reopen issue or create a new issue with different example if i missed something
